eo-phi-normalizer/src/Language/EO/Phi/Metrics/Collect.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedLabels #-}
+
+module Language.EO.Phi.Metrics.Collect where
+
+import Control.Lens ((+=))
+import Control.Monad (forM_)
+import Control.Monad.State (State, execState)
+import Data.Aeson (FromJSON)
+import Data.Generics.Labels ()
+import GHC.Generics (Generic)
+import Language.EO.Phi.Rules.Common ()
+import Language.EO.Phi.Syntax.Abs
+
+data Metrics = Metrics
+  { dataless :: Int
+  , applications :: Int
+  , formations :: Int
+  , dispatches :: Int
+  }
+  deriving (Generic, Show, FromJSON, Eq)
+
+defaultMetrics :: Metrics
+defaultMetrics =
+  Metrics
+    { dataless = 0
+    , applications = 0
+    , formations = 0
+    , dispatches = 0
+    }
+
+collectMetrics :: (Inspectable a) => a -> Metrics
+collectMetrics a = execState (inspect a) defaultMetrics
+
+class Inspectable a where
+  inspect :: a -> State Metrics ()
+
+instance Inspectable Program where
+  inspect (Program binding) = forM_ binding inspect
+
+instance Inspectable Binding where
+  inspect = \case
+    AlphaBinding attr obj -> do
+      inspect attr
+      inspect obj
+    EmptyBinding attr -> do
+      #dataless += 1
+      inspect attr
+    DeltaBinding _ -> pure ()
+    LambdaBinding _ -> #dataless += 1
+    MetaBindings _ -> pure ()
+
+instance Inspectable Attribute where
+  inspect = \case
+    Phi -> pure ()
+    Rho -> pure ()
+    Sigma -> pure ()
+    VTX -> pure ()
+    Label _ -> pure ()
+    Alpha _ -> pure ()
+    MetaAttr _ -> pure ()
```

Comment:
couldn't this have simply been
```suggestion
  inspect = const $ pure ()
```
?

Reply:
It could, but I wasn't sure it should have been.

Reply:
I still don't know what's a dataless attribute.